docs/cypress.md

```diff
@@ -0,0 +1,156 @@
+
+### Logging In
+This doesn't quite exist yet. Most tests will just want to start with the client in a 'logged in'
+state, so we should provide an easy way to start a test with element in this state. The
+`registrationSecret` provided when starting a Synapse can be used to create a user (porting
+the code from https://github.com/matrix-org/matrix-react-sdk/blob/develop/test/end-to-end-tests/src/rest/creator.ts#L49).
+We'd then need to log in as this user. Ways of doing this would be:
+
+1. Fill in the login form. This isn't ideal as it's effectively testing the login process in each
+   test, and will just be slower.
+1. Mint an access token using https://matrix-org.github.io/synapse/develop/admin_api/user_admin_api.html#login-as-a-user
+   then inject this into element-web. This would probably be fastest, although also relies on correctly
+   setting up localstorage
+1. Mint a login token, inject the Homeserver URL into localstorage and then load element, passing the login
+   token as a URL parameter. This is a supported way of logging in to element-web, but there's no API
+   on Synapse to make such a token currently. It would be fairly easy to add a synapse-specific admin API
+   to do so. We should write tests for token login (and the rest of SSO) at some point anyway though.
+
+If we make this as a convenience API, it can easily be swapped out later: we could start with option 1
+and then switch later.
+
+### Joining a Room
+Many tests will also want to start with the client in a room, ready to send & receive messages. Best
+way to do this may be to get an access token for the user and use this to create a room with the REST
+API before logging the user in.
+
+### Convenience APIs
+We should probably end up with convenience APIs that wrap the syanspe creation, logging in and room
+creation that can be called to set up tests.
+
+## Good Test Hygiene
+This section mostly summarises general good Cypress testing practice, and should not be news to anyone
+already familiar with Cypress.
+
+1. Test a well-isolated unit of functionality. The more specific, the easier it will be to tell what's
+   wrong when they fail. 
+1. Don't depend on state from other tests: any given test should be able to run in isolation.
+1. Try to avoid driving the UI for anything other than the UI you're trying to test. eg. if you're
+   testing that the user can send a reaction to a message, it's best to send a message using a REST
+   API, then react to it using the UI, rather than using the element-web UI to send the message.
+1. Avoid explicit waits. `cy.get()` will implicitly wait for the specified element to appear, so
+   you should never need to manually wait for an element. You can also wait for other things like
+   network requests in the browser to complete (https://docs.cypress.io/guides/guides/network-requests#Waiting).
+   Needing to wait for things can also be because of race conditions in the app itself, which ideally
+   shouldn't be there!
```

Comment:
What about for a test like editing an edited message, I don't think a selector will help us here because `(edited)` is already rendered, should we await on app-internal event emitters?

Reply:
Yeah, this is a good one - I've added it as an example with what I think would be the best way of doing it.